blog/2025-08-22-pattern.mdx

```diff
@@ -0,0 +1,312 @@
+most frequently for local variable assignments, allowing you to unpack values
+from expressions into new variables.
+
+## How to Destructure Values in Compact 🛠️
+
+Let's dive into the specifics of how to use tuple and structure patterns.
+The grammar for a pattern `p` is provided in
+[Compact's reference](https://docs.midnight.network/develop/reference/compact/lang-ref#parameters-patterns-and-destructuring).
+
+
+### Destructuring Tuple Values
+
+Tuple values can be destructured using a list of patterns enclosed in `[` and `]`.
+You can destructure a tuple into a sequence of identifiers:
+
+```compact
+const [a, b] = [1, true]; // 'a' is 1, 'b' is true
+```
+
+Sometimes you don't need all the values in a tuple. Compact allows you to skip
+elements by simply omitting the identifier but you cannot skip the actual element:
+
+```compact
+const [x, , z] = [10, 20, 30]; // 'x' is 10, 'z' is 30, 20 is skipped
+// const [x, , z] = [10, , 30]; // triggers a parse error
+```
+
+Skipped identifiers count towards the length of the pattern if there exists a
+named identifier after them. For example, `[x, y, ,]` has length 2 but
+`[x, , y]` has length 3.
+
+:::note
+If you want to enforce length 'l' of a tuple pattern you must have a named identifier
+at the `l-1` index of your pattern. For example, if you need to enforce a tuple
+pattern of legth 3 you need to have a named identifier on index 2 (e.g., all
+`[, , x]`, `[, x, y]`, and `[x, y, z]` enforce a tuple pattern of length 3).
+:::
+
+### Type Annotations for Tuple Patterns
+
+In a circuit definition, type declarations are required for tuple patterns:
+
+```compact
+circuit exampleTuplePattern([x, y]: [Field, Boolean]): Field {
+  // x has type Field, y has type Boolean
+  return x;
+}
+```
+
+Such a type declaration dictates the structure and the type of each element.  For example,
+in `exampleTuplePattern` the input has to be a tuple, it first element has to be a value
+of type `Field`, and its second element has to be a value of type `Boolean`.
+A call to `exampleTuplePattern` that does not meet these restrictions
+(e.g., `exampleTuplePattern([1, 1])`) will trigger a type error.
+
+For const bindings and anonymous circuits, type annotations are optional but recommended
+for clarity and an easier debugging experience. When a type annotation is provided
+ensure its structure aligns with the pattern. For example, `[a, , b] : [Field, Field, Field]`
+aligns the structure of the pattern and its type whereas `[a, , b] : [Field, Field]` does not:
+the pattern states that it is a tuple with three elements but the type annotation is
+tuple type of length two.
+
+Unlike patterns, types do not allow skipping elements in a tuple type. For example,
+`[a, , b] : [Field, , Field]` causes a parse error. So even when you're skipping an
+element in a pattern you must assign it a dummy type. Unfortunately, you have to
+remember what dummy type you used when you're passing arguments to pattern parameters.
```

Comment:
I don't think I would call this a "dummy type".  You have to pick a concrete type.

What you might want to do is make this circuit generic:

```
circuit second<T>([, x]: [T, Boolean]): Boolean {
  return x;
}
```
And then you have to specialize it correctly at the call site.